src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala

```diff
@@ -0,0 +1,283 @@
+  }
+
+  def saveWorkspace(dataAccess: DataAccess, updatedEntitiesAndWorkspace: Seq[Either[(Option[Entity], Option[Workspace]), (WorkflowRecord, scala.Seq[AttributeString])]]) = {
+    //note there is only 1 workspace (may be None if it is not updated) even though it may be updated multiple times so reduce it into 1 update
+    val workspaces = updatedEntitiesAndWorkspace.collect { case Left((_, Some(workspace))) => workspace }
+    if (workspaces.isEmpty) DBIO.successful(0)
+    else dataAccess.workspaceQuery.save(workspaces.reduce((a, b) => a.copy(attributes = a.attributes ++ b.attributes)))
+  }
+
+  def saveEntities(dataAccess: DataAccess, workspace: Workspace, updatedEntitiesAndWorkspace: Seq[Either[(Option[Entity], Option[Workspace]), (WorkflowRecord, scala.Seq[AttributeString])]]) = {
+    val entities = updatedEntitiesAndWorkspace.collect { case Left((Some(entity), _)) => entity }
+    if (entities.isEmpty) DBIO.successful(0)
+    else dataAccess.entityQuery.save(SlickWorkspaceContext(workspace), entities)
+  }
+
+  def attachOutputs(workspace: Workspace, workflowsWithOutputs: Seq[(WorkflowRecord, ExecutionServiceOutputs)], entitiesById: Map[UUID, Entity], outputExprepressions: Map[String, String]): Seq[Either[(Option[Entity], Option[Workspace]), (WorkflowRecord, Seq[AttributeString])]] = {
+    workflowsWithOutputs.map { case (workflowRecord, outputsResponse) =>
+      val outputs = outputsResponse.outputs
+
+      val attributes = outputExprepressions.map { case (outputName, attributeName) =>
+        Try {
+          attributeName.value -> outputs.getOrElse(outputName, {
+            throw new RawlsException(s"output named ${outputName} does not exist")
+          })
+        }
+      }
+
+      if (attributes.forall(_.isSuccess)) {
+        Left(updateEntityAndWorkspace(entitiesById(workflowRecord.workflowEntityId.get), workspace, attributes.map(_.get).toMap))
+
+      } else {
+        Right((workflowRecord, attributes.collect { case Failure(t) => AttributeString(t.getMessage) }.toSeq))
+      }
+    }
+  }
+
+  def updateEntityAndWorkspace(entity: Entity, workspace: Workspace, workflowOutputs: Map[String, Attribute]): (Option[Entity], Option[Workspace]) = {
+    //Partition outputs by whether their attributes are entity attributes (begin with "this.") or workspace ones (implicitly; begin with "workspace.")
+    //This assumption (that it's either "this." or "workspace.") will be guaranteed by checking of the method config when it's imported; see DSDEEPB-1603.
+    //Yes I know this is a var but it's more terse this way.
+    var (entityAttributes, workspaceAttributes) = workflowOutputs.partition({ case (k, v) => k.startsWith("this.") })
+    entityAttributes = entityAttributes.map({ case (k, v) => (k.stripPrefix("this."), v) })
+    workspaceAttributes = workspaceAttributes.map({ case (k, v) => (k.stripPrefix("workspace."), v) })
```

Comment:
this is actually copied from the old code, it works so I left it alone but if you want me to change it I will